Paint Brush documentation (#2099)
o3de/o3de.org: content/docs/user-guide/components/reference/paintbrush/paintbrush.md

---
linkTitle: Paint Brush
title: Paint Brush
description: ' The Paint Brush is a generalized tool for manipulating arbitrary data in world space using standard painting controls. '
---

In O3DE, many different components attach data to world space positions and regions. This data can be edited externally in content creation tools, but sometimes it can be easier to edit the data directly in the context of the assembled world inside of O3DE.

The Paint Brush is an editing mode available on some types of components to enable standard painting controls for editing the component's data in world space.

> Some components directly map pixels from a source image into world space, but other components might have a more indirect mapping, such as weights stored on vertices on a mesh. The Paint Brush is designed to allow painting in world space for any arbitrary type of data. For that reason, the descriptions below refer more generically to "values" instead of "pixels", even though painting is generally considered to be an image-based operation. Also, the mapping of the data values into world space can be at different densities, so the descriptions below make a distinction between "world space" and "value space". "World space" represents a number of meters in the virtual world. "Value space" represents a number of pixels / vertex weights / etc.

Author: How does this sound for the last sentence? "Value space represents a quantity of adjacent data values in the data source being edited."

Why did you remove "considered to be"? I'm not _opposed_ to removing it, but I thought it provided additional meaning. It's like the difference between "I thought the kitten was ugly, even though the kitten was generally considered to be cute" and "I thought the kitten was ugly, even though the kitten was generally cute." Since the sentence is challenging an assumption or expectation, I thought the phrase "considered to be" provided meaning, not fluff.

> Some components directly map pixels from a source image into world space, but other components might have a more indirect mapping, such as weights stored on vertices on a mesh. The Paint Brush is designed to allow painting in world space for any arbitrary type of data. For that reason, the descriptions below refer more generically to "values" instead of "pixels", even though painting is generally considered to be an image-based operation. Also, the mapping of the data values into world space can be at different densities, so the descriptions below make a distinction between "world space" and "value space". "World space" represents a number of meters in the virtual world. "Value space" represents a number of pixels / vertex weights / etc.

Reviewer: > How does this sound for the last sentence? "Value space represents a quantity of adjacent data values in the data source being edited."

Looks good!

Re: "Considered to be". I felt it was equivalent to generally; I agree there is a difference between the two.  "Considered to be" stresses the subjectivity of an evaluation, "generally" allows room for multiple evaluations while (subjectively?) preferring a default.  Keep your phrasing if you like it.

## Using the Paint Brush

The Paint Brush is enabled by clicking the **Edit** button on the component or by selecting the component's icon in the **Component Switcher** in the viewport.

Once the Paint Brush is active, there are three sets of controls:

![Paint Brush Tool image.](/images/user-guide/components/reference/paintbrush/paintbrush-tool.png)

1. A dockable **Paint Brush Settings** window
2. Paint Brush mode selection icons in the viewport
3. The Paint Brush circle in the viewport

Painting consists of moving the mouse in the viewport while holding down the left mouse button. A brush stroke begins once the left mouse button is pressed down and ends once the left mouse button is released. Undo/redo commands work on each brush stroke.

> A brush stroke consists of a series of overlapping circles with constant spacing that are applied as the mouse moves. The Paint Brush Settings control the color, opacity, size, and spacing of the circles.

Author: Hm. The sentence is supposed to say that the circles are applied, not that the spacing is applied. How about "A brush stroke consists of a series of overlapping circles applied with constant spacing as the mouse moves." ?

> A brush stroke consists of a series of overlapping circles with constant spacing that are applied as the mouse moves. The Paint Brush Settings control the color, opacity, size, and spacing of the circles.

Reviewer: We have the description in the previous sentence to rely on too.  Your new sentence works; you could also consider:

> A brush stroke applies a series of overlapping circles with constant spacing as the mouse moves.

or 

> As the mouse moves, a brush stroke applies a series of overlapping circles with constant spacing.

> A brush stroke consists of a series of overlapping circles with constant spacing that are applied as the mouse moves. The Paint Brush Settings control the color, opacity, size, and spacing of the circles.

Author: Hm. Those are both good suggestions, but they change the meaning a little bit to make it sound like the brush stroke _does_ a thing instead of the brush stroke _is_ a thing. I'll go with my new slightly wordier version.

![Example brush stroke.](/images/user-guide/components/reference/paintbrush/paintbrush-brush-stroke.png)

Painting mode is ended by pressing **Esc**, clicking the **Done** button on the component, or by selecting a different component's icon in the **Component Switcher** in the viewport.

## Paint Brush Modes

The Paint Brush has three separate modes - **Paint**, **Eyedropper**, and **Smooth**.

| Mode | Description |
| - | - |
| Paint | The Paint Brush blends new values onto whatever data is being painted based on the Paint Brush Settings. |

> | Eyedropper | The Paint Brush reads the current value from underneath the exact center of the brush and sets the Intensity / Color / Opacity in the Paint Brush Settings to that value. The Eyedropper always uses the exact center to read a single value, the brush size does not affect it. |

Author: Just for my edification, is this because the fall under the style guide for UI elements (bold every reference) as opposed to the style guide for Gems, components, applications, and tools (bold only the first reference)? If so, it would be nice to explicitly call out properties in the style guide.

> | Eyedropper | The Paint Brush reads the current value from underneath the exact center of the brush and sets the Intensity / Color / Opacity in the Paint Brush Settings to that value. The Eyedropper always uses the exact center to read a single value, the brush size does not affect it. |

Reviewer: Yes, it would fall under either UI elements or a new (undocumented) rule for properties explicitly.  It should be called out in the style guide, I'll create an issue for it.

| Overlay | $$a < 1/2 : c = 2ab$$<br>$$a >= 1/2 : c = 1 - 2(1 - a)(1 - b)$$ | If the main value is dark, apply Multiply mode and darken; if the main value is light, apply Screen mode and lighten. |

### Settings that affect brush circles

#### Size

Size controls the diameter of each brush circle. The size is specified in meters in world space coordinates, not in pixels or the number of values affected in the component.

#### Flow

**Flow** controls the maximum opacity of each brush circle. A Flow value of 0% is completely transparent, and 100% is completely opaque.

| Flow | Illustration |
| - | - |
| 1% | ![Paintbrush flow of 1%.](/images/user-guide/components/reference/paintbrush/paintbrush-flow-1.png) |
| 100% | ![Paintbrush flow of 100%.](/images/user-guide/components/reference/paintbrush/paintbrush-flow-100.png) |

Because the Flow opacity affects each brush circle, overlapping circles in the same brush stroke will blend with each other.

An example of a brush stroke crossing itself with medium flow:
![Example of brush stroke crossing itself with medium flow.](/images/user-guide/components/reference/paintbrush/paintbrush-flow-overlap.png)

#### Hardness

> **Hardness** controls the opacity falloff of each brush circle. The Hardness is the percent distance along the radius of the circle to begin the falloff. The falloff always ends at the outer radius. For example, a Hardness value of 50% means that the inner 50% of the brush circle will use the maximum opacity specified by Flow, and the outer 50% of the brush circle will drop in opacity until it is completely transparent. A value of 100% will make the entire brush circle use the maximum opacity.

Reviewer: ```suggestion
**Hardness** controls the opacity falloff of each brush circle. The **Hardness** is the percentage of distance along the circle's radius to begin the falloff at. The falloff always ends at the outer radius. For example, a **Hardness** value of 50% means that the inner 50% of the brush circle will use the maximum opacity specified by **Flow**, and the outer 50% of the brush circle will drop in opacity until it is completely transparent. A value of 100% will make the entire brush circle use the maximum opacity.
```

An illustration of Hardness 50% - the inner circle shows where the falloff begins, and the outer circle shows where the falloff ends. The outer circle is the brush size:![Illustration of hardness opacity falloff.](/images/user-guide/components/reference/paintbrush/paintbrush-falloff-50-percent.png)

> The following chart shows how the overall circle opacity changes through different combinations of Hardness and Flow. Although the circles appear to be different sizes, they are all the same brush size with differing amounts of opacity between the center and the outer edge. Lower Flow values have a lower maximum opacity which causes the falloff to reach 0 more quickly.

Reviewer: ```suggestion
The following chart shows how the overall circle opacity changes through different combinations of **Hardness** and **Flow**. Although the circles appear to be different sizes, they are all the same brush size with differing amounts of opacity between the center and the outer edge. Lower **Flow** values have a lower maximum opacity which causes the falloff to reach 0 more quickly.
```

| | Flow 1% | Flow 50% | Flow 100% |
| - | - | - | - |
| Hardness 0% | ![0% Hardness 1% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-0-flow-1.png) | ![0% Hardness 50% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-0-flow-50.png) | ![0% Hardness 100% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-0-flow-100.png) |
| Hardness 50% | ![50% Hardness 1% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-50-flow-1.png) | ![50% Hardness 50% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-50-flow-50.png) | ![50% Hardness 100% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-50-flow-100.png) |
| Hardness 100% | ![100% Hardness 1% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-100-flow-1.png) | ![100% Hardness 50% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-100-flow-50.png) | ![100% Hardness 100% Flow.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-100-flow-100.png) |

> Because the Hardness opacity affects each brush circle, overlapping circles in the same brush stroke will blend with each other.

Reviewer: ```suggestion
Because the **Hardness** opacity affects each brush circle, overlapping circles in the same brush stroke will blend with each other.
```

An example of a brush stroke crossing itself with medium hardness:![Example of brush stroke crossing itself with medium hardness.](/images/user-guide/components/reference/paintbrush/paintbrush-hardness-overlap.png)

#### Distance

> **Distance** controls the spacing of each circle within the brush stroke as the mouse moves. Very specifically, The Distance value represents the total distance the mouse needs to move before applying another circle, so small back-and-forth movements with the mouse can produce circles that are closer together than the Distance value. The Distance is a percentage of the circle's size, so a value of 50% will overlap each circle by 50% and a value of 100% will produce exactly non-overlapping circles. Values greater than 100% will leave space between each circle.

Reviewer: ```suggestion
**Distance** controls the spacing of each circle within the brush stroke as the mouse moves. Specifically, the **Distance** value represents the total distance the mouse needs to move before applying another circle, so small back-and-forth movements with the mouse can produce circles that are closer together than the **Distance** value. **Distance** is a percentage of the circle's size, so a value of 50% will overlap each circle by 50% and a value of 100% will produce exactly non-overlapping circles. Values greater than 100% will leave space between each circle.
```

| Distance | Illustration |
| - | - |
| 1% | ![Paintbrush distance of 1%.](/images/user-guide/components/reference/paintbrush/paintbrush-distance-1.png) |
| 25% | ![Paintbrush distance of 25%.](/images/user-guide/components/reference/paintbrush/paintbrush-distance-25.png) |
| 50% | ![Paintbrush distance of 50%.](/images/user-guide/components/reference/paintbrush/paintbrush-distance-50.png) |
| 100% | ![Paintbrush distance of 100%.](/images/user-guide/components/reference/paintbrush/paintbrush-distance-100.png) |
| 200% | ![Paintbrush distance of 200%.](/images/user-guide/components/reference/paintbrush/paintbrush-distance-200.png) |

### Settings that apply to smoothing operations

The following settings only apply to the smoothing brush.

#### Smooth Mode

**Smooth Mode** is the method for combining an NxN set of values together into a single smoothed value for use at the center of the NxN square.

| Mode | Description | Example 1 | Example 2 |
| - | - | - | - |

> | (None) | (Input images used in smoothing) |![Sample smoothing input image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-input.png) |![Sample smoothing input image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-input2.png) |

Reviewer: ```suggestion
| `None` | Input images used in smoothing. |![Sample smoothing input image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-input.png) |![Sample smoothing input image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-input2.png) |
```
If these are enumerated values of a property, they should use code style.  Should this description read "The input images use no smoothing."?

Author: So...maybe this table isn't doing what I had hoped. The "(None)" option _isn't_ an enumerated value. That row is here to provide the baseline of what things look like without smoothing, so that every image below can be compared to it to see how it changes with the different options. All of the other entries are the enumerated values.

Is there a better way to present this information that would be less confusing?

Reviewer: OK, I didn't realize that.  At the very least, you can drop 'None' entirely and leave that cell of the table empty.  I'm fine with either your original or my suggestion for the Description cell.  Maybe just (Input images)?  Your call.

Author: Oooh, good idea. I'll drop the 'None' and change the Description to '(Input images)'.

> | Gaussian | Takes a weighted average of the NxN values using a Gaussian bell curve distribution function. This is the most common smoothing method used by paint programs. It provides a good balance of smoothing out noise and preserving details. |![Sample smoothing Gaussian image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-gaussian.png) |![Sample smoothing Gaussian image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-gaussian2.png) |

Reviewer: ```suggestion
| `Gaussian` | Takes a weighted average of the NxN values using a Gaussian bell curve distribution function. This is the most common smoothing method used by paint programs. It provides a good balance of smoothing out noise and preserving details. |![Sample smoothing Gaussian image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-gaussian.png) |![Sample smoothing Gaussian image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-gaussian2.png) |
```

> | Mean | Takes an unweighted average of the NxN values. This method smooths out noise and is the method least likely to preserve sharp details. |![Sample smoothing mean image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-mean.png) |![Sample smoothing mean image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-mean2.png) |

Reviewer: ```suggestion
| `Mean` | Takes an unweighted average of the NxN values. This method smooths out noise and is the method least likely to preserve sharp details. |![Sample smoothing mean image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-mean.png) |![Sample smoothing mean image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-mean2.png) |
```

> | Median | Takes the middle value from the sorted set of NxN values. This method preserves sharper edges than the other smoothing methods and can completely eliminate noise, but it also produces less smooth results in general and might introduce noticeable artifacts. |![Sample smoothing median image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-median.png) |![Sample smoothing median image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-median2.png) |

Reviewer: ```suggestion
| `Median` | Takes the middle value from the sorted set of NxN values. This method preserves sharper edges than the other smoothing methods and can completely eliminate noise, but it also produces less smooth results in general and might introduce noticeable artifacts. |![Sample smoothing median image.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-median.png) |![Sample smoothing median image 2.](/images/user-guide/components/reference/paintbrush/paintbrush-smooth-mode-median2.png) |
```

#### Smoothing Radius

> **Smoothing Radius** is the number of values to use in each direction from the center value when computing the smoothed value. The smoothing methods use NxN values, where `N = (radius * 2) + 1`. A radius of 1 is a 3x3 set of values, a radius of 2 is a 5x5 set of values, a radius of 3 is a 7x7 set of values, etc.

Reviewer: ```suggestion
**Smoothing Radius** is the number of values to use in each direction from the center value when computing the smoothed value. The smoothing methods use NxN values, where `N = (radius * 2) + 1`. A radius of 1 is a 3x3 set of values, a radius of 2 is a 5x5 set of values, a radius of 3 is a 7x7 set of values, and so on.
```

> This number is expressed in "Value Space". In other words, it is describing the number of adjacent values to use in each direction regardless of how the values are mapped into world space.

Reviewer: ```suggestion
This number is expressed in *value space*. In other words, it describes the number of adjacent values to use in each direction regardless of how the values are mapped into world space.
```
Here, the definition of value space is reiterated.  We can use italics again.

> A larger smoothing radius will generally cause a smoother result across larger distances, but it also has an increased performance cost.

Reviewer: ```suggestion
A larger radius will generally cause a smoother result across larger distances, but it also has an increased performance cost.
```

> The Smoothing Radius is separate from the Paint Brush Size. The Paint Brush Size controls how many values will get smoothed, and the Smoothing Radius controls how many adjacent values are used to smooth each value. Only values within the Paint Brush will get smoothed, but values outside of the Paint Brush might be used for calculating the smoothed results.

Reviewer: ```suggestion
The **Smoothing Radius** is separate from the Paint Brush **Size**. The Paint Brush **Size** controls how many values will get smoothed, and the **Smoothing Radius** controls how many adjacent values are used to smooth each value. Only values within the Paint Brush will get smoothed, but values outside of the Paint Brush might be used for calculating the smoothed results.
```

## Hotkeys and CVars

| Hotkey | Description |
| - | - |

> | `[` | Decreases the Paint Brush Size |

Reviewer: ```suggestion
| **[** | Decreases the Paint Brush **Size**. |
```
1. Hotkeys are bolded.
2. Sidenote: This is an example of where bolding properties really shines; the user can be assured that this is going to affect the property's value directly.

> | `]` | Increases the Paint Brush Size |

Reviewer: ```suggestion
| **]** | Increases the Paint Brush **Size**. |
```

> | `{` | Decreases the Paint Brush Hardness |

Reviewer: ```suggestion
| **{** | Decreases the Paint Brush **Hardness**. |
```

> | `}` | Increases the Paint Brush Hardness |

Reviewer: ```suggestion
| **}** | Increases the Paint Brush **Hardness**. |
```
